Create a post with a quick farmOS.py tutorial

symbioquine opened the pull request.

mstenta commented on posts/2023/quick-intro-to-python-scripting/index.md:
```diff
@@ -0,0 +1,233 @@
+# Introduction to Python Scripting with farmOS.py
+
+Although farmOS has a convenient web interface that is useful for record keeping and reference tasks by humans, sometimes it is
+necessary to programmatically read or write farmOS data. One of the libraries that facilitates that is farmOS.py, the official
```
Link "farmOS.py" to https://github.com/farmOS/farmOS.py/ ?

mstenta commented on posts/2023/quick-intro-to-python-scripting/index.md:
```diff
@@ -0,0 +1,233 @@
+
+# Introduction to Python Scripting with farmOS.py
+
+Although farmOS has a convenient web interface that is useful for record keeping and reference tasks by humans, sometimes it is
```
I don't think we've officially established this as a standard for blog posts, but in most of our other repos we've standardized on a character limit of 80 in markdown files.

symbioquine replied:
Reformatted and added a [`.prettierrc` config file](https://prettier.io/docs/en/configuration.html) to the repo.

paul121 commented on posts/2023/quick-intro-to-python-scripting/index.md:
````diff
@@ -0,0 +1,233 @@
+
+```shell
+export REQUESTS_CA_BUNDLE=/path/to/your/rootCA.pem
+```
````
This would be nice to add to the official docs :-)

symbioquine replied:
Probably not personally taking that on right now... while we're at it though, I think [this line](https://github.com/farmOS/farmOS.py/blame/8246dcae86479f82a57c8d90336ed738622cca34/docs/client_2x.md#L220) in the docs is wrong also;

![image](https://user-images.githubusercontent.com/30754460/224714844-19e31069-baba-4f2c-9e74-eedca8a8ac00.png)

`'plant'` should be `'plant_type'`

paul121 commented on posts/2023/quick-intro-to-python-scripting/index.md:
````diff
@@ -0,0 +1,233 @@
+poetry run python harvest_and_archive_last_season_crops.py
+```
+
+Should print something like;
+
+```
+Harvesting and archiving: 2022 L12 Radish: 'French Breakfast'
+Harvesting and archiving: 2022 L8 Potato: Ozette
+Harvesting and archiving: 2022 L8 Potato: Purple Fiesta
````
A nice addition might be to generate + print a link to the harvest log in farmOS, what do you think?

symbioquine replied:
I'll give it a shot and see if the added code makes the example too complicated or not...

paul121 commented on posts/2023/quick-intro-to-python-scripting/index.md:
````diff
@@ -0,0 +1,233 @@
+```shell
+mkdir farm_python_scripting_tutorial && cd farm_python_scripting_tutorial
+poetry init
+poetry add "farmOS^1.0.0b"
````
I would like to tag a stable release of farmOS.py v1 before we post this! 
I've been holding off for a stable farmOS v2 release but no longer have that excuse... I will try and do this tomorrow :smile: 

People have had trouble with the beta tags, either missing or using the wrong syntax. You have it correct here but it would just be cleaner and better for the future. In fact maybe it is best to use `farmOS^1` just in case a v2 does come along and no longer works with this blog post...

symbioquine replied:
Awesome! Totally makes sense. I'll wait for a stable release to be tagged.

paul121 replied:
Done!

paul121 reviewed (commented):
This is great @symbioquine !!

I wish I could get a better markdown preview but a couple thoughts... what do you think about including the shell command + output blocks before the python source block? I got a little lost while reviewing which is why I mention.

I also think we may be able to remove the "Which prints something like:..." text and combine the `poetry run python ...`  and output together in one block. Having this above the code seems like a nice teaser + intro to what the code is doing?

Last, I think it would be great if we could include a super simple CSV import example here. I think that is a common use of farmOS.py scripting, could certainly be its own blog post, but maybe worth a little teaser here? I'd be happy to add an example!

symbioquine commented:
> This is great @symbioquine !!

Thanks, and thank you for all the great feedback!

> I wish I could get a better markdown preview but a couple thoughts... what do you think about including the shell command + output blocks before the python source block? I got a little lost while reviewing which is why I mention.
> 
> I also think we may be able to remove the "Which prints something like:..." text and combine the `poetry run python ...` and output together in one block. Having this above the code seems like a nice teaser + intro to what the code is doing?

I'd definitely welcome any feedback on the presentation! Here's a screenshot of it rendered if that helps validate/change your feedback above...

![image](https://user-images.githubusercontent.com/30754460/224197922-d3d21b37-c145-4f7c-b473-6a4f4a125dae.png)

> Last, I think it would be great if we could include a super simple CSV import example here. I think that is a common use of farmOS.py scripting, could certainly be its own blog post, but maybe worth a little teaser here? I'd be happy to add an example!

I like that idea! Unless you have one ready to go, I'll probably rip-off/adapt [my example](https://gist.github.com/symbioquine/7641a2ab258726347ec937e8ea02a167) from the [JupyterLite thread](https://farmos.discourse.group/t/share-your-jupyterlite-examples/1134/2#animal-csv-import-2).

mstenta reviewed (commented):
This is great @symbioquine!!! Makes me want to do more with farmOS.py myself! :smile: 

Only two minor comments on formatting...

paul121 commented:
> I like that idea! Unless you have one ready to go, I'll probably rip-off/adapt [my example](https://gist.github.com/symbioquine/7641a2ab258726347ec937e8ea02a167) from the [JupyterLite thread](https://farmos.discourse.group/t/share-your-jupyterlite-examples/1134/2#animal-csv-import-2).

Ooh. Importing the animals looks great. I see this is using `pandas.read_csv()`. I'm not familiar but I imagine this provides many helpers that is preferable over reading the CSV into a dict. Here is the approach I have used in the past (not a simple example but just to demonstrate): https://gist.github.com/paul121/4e048ec346ea17ce45a4df6f43a31295#file-harvest_import-py-L114-L118

I don't have much of an opinion on the CSV library, more just an observation and note that I'd like to learn more about Pandas :-)

paul121 commented:
> I'd definitely welcome any feedback on the presentation! Here's a screenshot of it rendered if that helps validate/change your feedback above...

Thank you! I think moving the source code below would be nice. Or at least add a more descriptive (short) header for each example. The python file name describes it but with the code highlighting it is confusing. Combining the output code block seems less important though

symbioquine commented:
I think I've addressed all the feedback above @paul121 / @mstenta . Here is the rendered preview;

https://user-images.githubusercontent.com/30754460/224712826-a6b90fb0-2534-4e4a-845c-7b083cc30ef7.png

> Ooh. Importing the animals looks great. I see this is using `pandas.read_csv()`. I'm not familiar but I imagine this provides many helpers that is preferable over reading the CSV into a dict. Here is the approach I have used in the past (not a simple example but just to demonstrate): https://gist.github.com/paul121/4e048ec346ea17ce45a4df6f43a31295#file-harvest_import-py-L114-L118
> 
> I don't have much of an opinion on the CSV library, more just an observation and note that I'd like to learn more about Pandas :-)

Yep, that's the same way I'd do the actual CSV import part in plain Python too. No reason to use Pandas for this post since that's just an extra thing for folks to learn/install. I only did it in the JupyterLite example because it provides nice table formatting there and it is the conventional way of doing it in that environment.

paul121 reviewed (approved):
Awesome!
